### ISSUE TITLE
Optimizer window pagination

### ISSUE THREAD

**Issue by Kitefiko** [strawberry-graphql/strawberry-django#613]
## Describe the Bug
I believe I have found multiple issues with window pagination implementation.
- `AttributeError` is always thrown in `ListConnectionWithTotalCount.resolve_connection_from_cache`
- backup solution in case of `AttributeError` prevents pagination
- order and thus pagination itself is non-deterministic when explicit order is not used
- usage of `last` returns everything and bypasses `max_results`

## System Information

 - Operating system: linux
 - Strawberry version (if applicable): 0.47.1
 - Database version: 10.11.6-MariaDB

### `ListConnectionWithTotalCount.resolve_connection_from_cache`
`AttributeError` is always thrown here:
https://github.com/strawberry-graphql/strawberry-django/blob/d6051db6f1e792604d66206734df9d6ebcfbbb40/strawberry_django/relay.py#L195
I think there should be something like this?
```python
has_next_page = result[-1]._strawberry_row_number < result[-1]._strawberry_total_count if result else False
```

### Backup solution in case of `AttributeError` prevents pagination
This _might_ be intended behavior, but:
In case of `AttributeError` pagination falls back to `resolve_connection` but (probably due to unavailable overfetch?) `pageInfo`'s `hasNextPage` is wrongly reporting as `false`

### Order and thus pagination itself is non-deterministic when explicit order is not used
This is probably the biggest issue. Maybe fixable via `pk` ordering in case of empty `order_by` in `apply_window_pagination` ?

### Usage of `last` returns everything and bypasses `max_results`
When `last` is used on it's own and execution ends up here:
https://github.com/strawberry-graphql/strawberry-django/blob/d6051db6f1e792604d66206734df9d6ebcfbbb40/strawberry_django/optimizer.py#L561-L573

`slice_metadata.start` is evaluated to 0
`slice_metadata.end` becomes `sys.maxsize`

and queryset ends up filtering out nothing

**Comment by bellini666** [comment]
Hey @Kitefiko ,

Hrm, I'm not really I understood the issue. When and why does `AttributeError` gets raised for you? Could you try to provide an MRE for it?

**Comment by SupImDos** [comment]
Hi @bellini666

We've run into these issue as well, particularly with `ListConnectionWithTotalCount.resolve_connection_from_cache` here:

https://github.com/strawberry-graphql/strawberry-django/blob/d6051db6f1e792604d66206734df9d6ebcfbbb40/strawberry_django/relay.py#L195

Just so its clear, the `result` object here is actually the `QuerySet._result_cache` `list`, so it will never have the annotated `_strawberry_row_number` / `_strawberry_total_count` attributes. It's items will though, so as suggested in parent issue it should instead be something like:

```python
has_next_page = (
    result[-1]._strawberry_row_number < result[-1]._strawberry_total_count
    if result
    else False
)
```

**Comment by SupImDos** [comment]
@bellini666 FYI #622 only fixes 1 of the 4 problems raised in this issue, so it may be worth keeping this open?

**Comment by bellini666** [comment]
@SupImDos good point, going to reopen this

**Comment by bellini666** [comment]
@SupImDos @Kitefiko could any of you help me with some MREs for the remaining issues here so I can try to tackle?

**Comment by SupImDos** [comment]
> [@SupImDos](https://github.com/SupImDos) [@Kitefiko](https://github.com/Kitefiko) could any of you help me with some MREs for the remaining issues here so I can try to tackle?

@bellini666 Yes, we've specifically run into the _"Usage of last returns everything and bypasses max_results"_ issue, so I should be able to put together an MRE for that soon.

**Comment by SupImDos** [comment]
@bellini666 
Sorry for the delay! I've created an MRE for the _"Usage of last returns everything and bypasses max_results"_ issue:
* Branch: [`mre/pagination-with-only-last`](https://github.com/BrightpathProgress/strawberry-django/tree/mre/pagination-with-only-last)
* Commit: 922ec47b8a7fd2e48a48b8b6cdc6f385804e3254

In short, doing:
```graphql
query {
  milestoneConn(last: 3) {
    edges {
      node {
        id
      }
    }
  }
}
```

Results in:
```sql
SELECT "projects_milestone"."name",
    "projects_milestone"."id",
    "projects_milestone"."due_date",
    "projects_milestone"."project_id"
FROM "projects_milestone" LIMIT 9223372036854775807    # <- This is ``sys.maxsize``
```

**Comment by bellini666** [comment]
Hi @SupImDos ,

Thanks for the MRE! 😊

Here is what is happening: The relay limit/offset algorithm on strawberry is implemented using slices, and those are calculated here: https://github.com/strawberry-graphql/strawberry/blob/main/strawberry/relay/utils.py#L126

If you take a look at [this line](https://github.com/strawberry-graphql/strawberry/blob/main/strawberry/relay/utils.py#L167), you will see that if you only ask for `last` without anything else, it will do a slice like `nodes[: sys.maxint]` and then indeed get the last results in python.

A good workaround would be to make this do `nodes[-last:]`, but when I tried that, django raised `ValueError: Negative indexing is not supported.`

I think the only options we have to solve this specific case (we only have last and nothing else) would be:

1) Reverse the sorting in the queryset and change `last` to `first`

In theory, this would work the same, but then we would need to re-reverse the results after.

2) Call `.count()` in the queryset before the slice so we can do `nodes[count - last:]`

This one would not require re-reversing, but is somewhat more complicated to implement, as it would require changing the strawberry API to allow hooking into it.


What do you think? Anything that I'm missing here?

**Comment by Kitefiko** [comment]
Hi @bellini666,
sorry for the delay. I am planing to look into this this week (finally).

The issues that I plan to address, that are connected do optimizer, are:

#### Using `last` without `before` & `after` - root (not nested) connection
As you mentioned above there are two options (a) using `.count()` (b) reverse order, slice as usual, reverse back

I believe that using `.count()`, adjusting slice_metadata and adjusting `ListConnectionWithTotalCount.resolve_connection` accordingly, would be the best approach.

Approach via ordering and reversing would require to _always_ use ordering, because non-ordered (natural) order might not match reverse order via specific field (id for example). I don't know this for sure tho.

#### There must be order - nested connection using `WINDOW`
I do not know if this is behavior of specific DBs only but, when using `OVER` without `order_by`  the generated rows in each `partition` do not always have the same order, thus do not have the same `RowNumber` and the entire pagination is useless.

This can be fixed by always ordering at least via `id` (when user is not ordering directly)

#### Using `last` without `before` & `after` - nested connection using `WINDOW`
This one is a bit more tricky. But query would have to be adjusted to fetch `.count` for each `partition`, then adjust `slice_metadata` and then query as usual. OR doing reverse trick since `There must be order (issue above)`?

#### Optimizing total_count - nested connection using `WINDOW`
"Overfetching" in `window_pagination` filter so that first row is always included. This would guarantee source of `total_count`. - This is bonus, not a issue. Additionally even non-nested connections _could_ be optimized via `annotate`, right?

I will start working on solution using `.count()`. What do you think?

**Comment by bellini666** [comment]
@Kitefiko sorry for not replying earlier, but go ahead with that and I'll gladly review the PR once it is ready :)

**Comment by bellini666** [comment]
Fixed by #777
